File: JPP/Lista4/Go/philosophers.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"
)

var (
	n         int
	forks     []chan bool
	waitGroup sync.WaitGroup
	meals     int
)

func validateArg(arg string, name string) (int, error) {
	val, err := strconv.Atoi(arg)
	if err != nil {
		return 0, fmt.Errorf("Error: %s is not an integer", name)
	}
	return val, nil
}

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Correct usage: go run philosophers.go philosophers_count meals_to_eat")
		return
	}

	var err error
	if n, err = validateArg(os.Args[1], "n"); err != nil {
		fmt.Println(err)
		return
	}

	if meals, err = validateArg(os.Args[2], "meals"); err != nil {
		fmt.Println(err)
		return
	}

	if n < 2 {
		fmt.Println("Error: One fork per Philosopher, at least 2 Philosophers required")
		return
	}

	if meals == 0 {
		fmt.Println("Error: Philosophers are hungry! They need to eat at least one meal")
		return
	}

	forks = make([]chan bool, n)
	for i := 0; i < n; i++ {
		forks[i] = make(chan bool, 1)
	}

	for i := 0; i < n; i++ {
		waitGroup.Add(1)
		go HandlePhilosopher(&waitGroup, i)
	}

	waitGroup.Wait()
}

func HandlePhilosopher(waitGroup *sync.WaitGroup, i int) {
	defer waitGroup.Done()

	for meal := 0; meal < meals; meal++ {
		// Thinking
		thinkingTime := time.Duration(rand.Intn(801) + 200)
		time.Sleep(thinkingTime * time.Millisecond)

		// Pick up Forks but consider parity of i variable
		if i%2 == 0 {
			fmt.Printf("[%d] is picking up fork %d\n", i, i)
			forks[i] <- true
			fmt.Printf("[%d] is picking up fork %d\n", i, (i+1)%n)
			forks[(i+1)%n] <- true
		} else {
			fmt.Printf("[%d] is picking up fork %d\n", i, (i+1)%n)
			forks[(i+1)%n] <- true
			fmt.Printf("[%d] is picking up fork %d\n", i, i)
			forks[i] <- true
		}

		// Eating
		fmt.Printf("    [%d] starts eating\n", i)
		eatingTime := time.Duration(rand.Intn(801) + 200)
		time.Sleep(eatingTime * time.Millisecond)

		// Put down Forks
		fmt.Printf("  [%d] is putting down fork %d\n", i, i)
		<-forks[i]
		fmt.Printf("  [%d] is putting down fork %d\n", i, (i+1)%n)
		<-forks[(i+1)%n]
	}
}

File: JPP/Lista4/Third/philosophers.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <chrono>
#include <vector>
#include <random>

class DiningPhilosophers {
public:
    DiningPhilosophers(int n, int meals) : n(n), meals(meals), forks(n), waitGroup(n) {}

    void start() {
        std::vector<std::thread> threads;
        for (int i = 0; i < n; ++i) {
            threads.emplace_back(&DiningPhilosophers::handlePhilosopher, this, i);
        }
        for (auto& t : threads) {
            t.join();
        }
    }

private:
    int n;
    int meals;
    std::vector<std::mutex> forks;
    std::vector<std::thread::id> waitGroup;
    std::mutex printMutex; // Mutex for protecting standard output

    void handlePhilosopher(int i) {
        for (int meal = 0; meal < meals; ++meal) {
            // // Thinking
            // std::this_thread::sleep_for(std::chrono::milliseconds(rand() % 1000 + 200));

            // Pick up forks
            std::unique_lock<std::mutex> lock1(forks[i]);
            {
                std::lock_guard<std::mutex> guard(printMutex); // Lock before printing
                std::cout << "[" << i << "]" << " picked up fork " << i << std::endl;
            }
            std::unique_lock<std::mutex> lock2(forks[(i + 1) % n]);
            {
                std::lock_guard<std::mutex> guard(printMutex); // Lock before printing
                std::cout << "[" << i << "]" << " picked up fork " << (i + 1) % n << std::endl;
            }

            // Eating
            {
                std::lock_guard<std::mutex> guard(printMutex); // Lock before printing
                std::cout << "    [" << i << "]" << " starts eating" << std::endl;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(rand() % 1000 + 200));

            // Put down forks
            lock1.unlock();
            std::cout << "  [" << i << "]" << " put down fork " << i << std::endl;
            lock2.unlock();
            std::cout << "  [" << i << "]" << " put down fork " << (i + 1) % n << std::endl;

        }
    }
};

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Correct usage: ./dpProblem_asymmetric philosophers_count meals_to_eat" << std::endl;
        return 1;
    }

    int n = std::stoi(argv[1]);
    int meals = std::stoi(argv[2]);

    if (n < 2) {
        std::cerr << "Error: One fork per Philosopher, at least 2 Philosophers required" << std::endl;
        return 1;
    }

    if (meals == 0) {
        std::cerr << "Error: Philosophers are hungry! They need to eat at least one meal" << std::endl;
        return 1;
    }

    srand(time(NULL));
    DiningPhilosophers dp(n, meals);
    dp.start();

    return 0;
}
